From 1f34fac2b65f1d008017ab5b4e9807b8581e1282 Mon Sep 17 00:00:00 2001
From: Vijay Mali <vmali@nvidia.com>
Date: Wed, 11 Jul 2012 14:06:23 +0530
Subject: [PATCH] external: tinyalsa: Enable tinyalsa apps in JB build

Enable tinyalsa apps in default JB builds. These apps are
required for audio sanity validation.

For bug 1014630
For bug 846501

Change-Id: I8fa43b457db6cc16c30bb50d15ede148f556a69a
Signed-off-by: Vijay Mali <vmali@nvidia.com>

tinyalsa: return card number from its name substring

tinyalsa doesn't provide any way to get card number from its name.
Added function to return card number from card name substring. This
is needed as usb audio device can be plug-in and plug-out anytime
which changes card number.

Bug 1009921
Bug 1026047

Change-Id: I2191d5eed94b030bf7fdca5f3e6a50d17b72113b

tinyalsa: search driver and long name for substring

Some usb audio devices doesn't have usb substring in their name and
id. So try matching driver name and longname.

Bug 1038505

Change-Id: I65da0d931f40bc4fa62e9b1165c3868d9fb18881

tiny_alsa: suspend state handling in pcm write

The device suspend error is handeled separately in
pcm write.

Bug 1030598

Change-Id: Iccaeaae4caaa0622edb6e3e30f5273afadeffa08
---
 Android.mk                   |    2 +-
 include/tinyalsa/asoundlib.h |    3 +++
 mixer.c                      |   40 ++++++++++++++++++++++++++++++++++++++++
 pcm.c                        |    2 ++
 4 files changed, 46 insertions(+), 1 deletions(-)

diff --git a/Android.mk b/Android.mk
index 5b64934..72f9054 100644
--- a/Android.mk
+++ b/Android.mk
@@ -6,7 +6,7 @@ LOCAL_SRC_FILES:= mixer.c pcm.c
 LOCAL_MODULE := libtinyalsa
 LOCAL_SHARED_LIBRARIES:= libcutils libutils
 LOCAL_MODULE_TAGS := optional
-LOCAL_PRELINK_MODULE := false
+LOCAL_PRELINK_MODULE := true
 
 include $(BUILD_SHARED_LIBRARY)
 
diff --git a/include/tinyalsa/asoundlib.h b/include/tinyalsa/asoundlib.h
index a796a66..63c97d6 100644
--- a/include/tinyalsa/asoundlib.h
+++ b/include/tinyalsa/asoundlib.h
@@ -205,6 +205,9 @@ int mixer_ctl_set_enum_by_string(struct mixer_ctl *ctl, const char *string);
 int mixer_ctl_get_range_min(struct mixer_ctl *ctl);
 int mixer_ctl_get_range_max(struct mixer_ctl *ctl);
 
+/* Get sound card number using sub string */
+int snd_card_get_card_id_from_sub_string(const char *card_name);
+
 #if defined(__cplusplus)
 }  /* extern "C" */
 #endif
diff --git a/mixer.c b/mixer.c
index 9514528..6b37dcf 100644
--- a/mixer.c
+++ b/mixer.c
@@ -416,3 +416,43 @@ int mixer_ctl_set_enum_by_string(struct mixer_ctl *ctl, const char *string)
     return -EINVAL;
 }
 
+
+int snd_card_get_card_id_from_sub_string(const char *card_name)
+{
+    int card;
+    int fd;
+    char fn[256];
+    struct snd_ctl_card_info info;
+
+    if (card_name == NULL)
+        return -EINVAL;
+
+    for (card = 0; card < 8; card++) {
+        snprintf(fn, sizeof(fn), "/dev/snd/controlC%u", card);
+        fd = open(fn, O_RDONLY);
+        if (fd < 0)
+            continue;
+        if (ioctl(fd, SNDRV_CTL_IOCTL_CARD_INFO, &info) < 0) {
+            close(fd);
+            continue;
+        }
+        close(fd);
+
+        /* search by card id */
+        if (strcasestr((const char *)info.id, card_name))
+            return info.card;
+
+        /* search by card name */
+        if (strcasestr((const char *)info.name, card_name))
+            return info.card;
+
+        /* search by driver name */
+        if (strcasestr((const char *)info.driver, card_name))
+            return info.card;
+
+        /* search by longname name */
+        if (strcasestr((const char *)info.longname, card_name))
+            return info.card;
+    }
+    return -EINVAL;
+}
diff --git a/pcm.c b/pcm.c
index 284a7ac..2b2e1bb 100644
--- a/pcm.c
+++ b/pcm.c
@@ -395,6 +395,8 @@ int pcm_write(struct pcm *pcm, const void *data, unsigned int count)
                 if (pcm->flags & PCM_NORESTART)
                     return -EPIPE;
                 continue;
+            } else if (errno == ESTRPIPE) {
+                    return -ESTRPIPE;
             }
             return oops(pcm, errno, "cannot write stream data");
         }
-- 
1.7.5.4

